test/e2e/ipam/ipamhelper.go

```diff
@@ -0,0 +1,373 @@
+	if err != nil {
+		return nil, err
+	}
+
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return client.New(restConfig, client.Options{Scheme: ipamScheme})
+}
+
+func (s *ipamService) claimIPAddress(ctx context.Context) (_ string, err error) {
+	claim := &ipamv1.IPAddressClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "ipclaim-" + rand.String(32),
+			Namespace: metav1.NamespaceDefault,
+			Labels:    getIPAddressClaimLabels(),
+		},
+		Spec: ipamv1.IPAddressClaimSpec{
+			PoolRef: corev1.TypedLocalObjectReference{
+				APIGroup: ptr.To("ipam.cluster.x-k8s.io"),
+				Kind:     "InClusterIPPool",
+				Name:     "capv-e2e-ippool",
+			},
+		},
+	}
+
+	// Create an IPAddressClaim
+	Byf("Creating IPAddressClaim %s/%s", claim.GetNamespace(), claim.GetName())
+	if err := s.client.Create(ctx, claim); err != nil {
+		return "", err
+	}
+	// Store claim inside the service so the cleanup function knows what to delete.
+	s.ipAddressClaims = append(s.ipAddressClaims, *claim)
+	ip := &ipamv1.IPAddress{}
+
+	var retryError error
+	// Wait for the IPAddressClaim to refer an IPAddress.
+	_ = wait.PollUntilContextTimeout(ctx, time.Second, time.Second*30, true, func(ctx context.Context) (done bool, err error) {
+		if err := s.client.Get(ctx, client.ObjectKeyFromObject(claim), claim); err != nil {
+			retryError = errors.Wrap(err, "getting IPAddressClaim")
+			return false, nil
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			retryError = errors.Wrap(err, "IPAddressClaim.Status.AddressRef.Name is not set")
+			return false, nil
+		}
+
+		if err := s.client.Get(ctx, client.ObjectKey{Namespace: claim.GetNamespace(), Name: claim.Status.AddressRef.Name}, ip); err != nil {
+			retryError = errors.Wrap(err, "getting IPAddress")
+			return false, nil
+		}
+		if ip.Spec.Address == "" {
+			retryError = errors.Wrap(err, "IPAddress.Spec.Address is not set")
+			return false, nil
+		}
+
+		retryError = nil
+		return true, nil
+	})
+	if retryError != nil {
+		return "", retryError
+	}
+
+	return ip.Spec.Address, nil
+}
+
+// cleanup deletes the IPAddressClaims created by the service if the test succeeded.
+func (s *ipamService) cleanup() {
+	if s.skipCleanup {
+		By("Skipping cleanup of IPAddressClaims because skipCleanup is set")
+		return
+	}
+	if CurrentSpecReport().Failed() {
+		By("Skipping cleanup of IPAddressClaims because the tests failed and the IPs could still be in use")
+		return
+	}
+
+	var errList []error
+
+	for _, ipAddressClaim := range s.ipAddressClaims {
+		ipAddressClaim := ipAddressClaim
+		Byf("Deleting IPAddressClaim %s/%s", ipAddressClaim.GetNamespace(), ipAddressClaim.GetName())
+		if err := s.client.Delete(context.Background(), &ipAddressClaim, client.DryRunAll); err != nil {
+			errList = append(errList, err)
+		}
+	}
+
+	if len(errList) > 0 {
+		Expect(kerrors.NewAggregate(errList)).ToNot(HaveOccurred())
+	}
+}
+
+func getIPAddressClaimLabels() map[string]string {
```

Comment:
I wonder if we have to safeguard against this returning an empty map.

In this case we would create IPAddressClaims without labels and then try to delete all of them (even the ones unrelated to our test run) (+/- only if corresponding VMs don't exist)

Reply:
If I'm not mistaken you are only adding the label key when the env var is set

Reply:
ah damn, I should have looked at the code first. You're totally right.
My thinking was: when running it locally that you don't have to set them.